### ISSUE TITLE
memory leak in tensorflow_gpu 0.12.1

### ISSUE THREAD

**Issue by Mahdizade** [tensorflow/tensorflow#6599]
**Hi, We use tensorflow for training our OCR system models. I simply train models in tensorflow 0.9 and former versions. after some upgrade in cuda and tensorflow, I see large memory leak in our server with 32GB RAM.**

### I have tried all of suggestions in [How to debug a memory leak in TensorFlow](http://stackoverflow.com/documentation/tensorflow/3883/how-to-debug-a-memory-leak-in-tensorflow#t=201612280142239281993) and some other github issues and stackoverflow posts. No of them worked.

train code:

```
#!/bin/env python
import tensorflow as tf

import Config
import Utilities
from Dataset import Dataset
from Model import Model

dataset = Dataset()

sess_config = tf.ConfigProto()
sess_config.gpu_options.allow_growth = True
sess = tf.Session(config=sess_config)
images, labels = dataset.train_images_labels()
model = Model(images, labels, training=True)
tf.train.start_queue_runners(sess=sess)


def main(global_step=0):
    if global_step == 0:
        init_op = tf.global_variables_initializer()
        sess.run(init_op)
    else:
        checkpoint_path = Utilities.get_checkpoint_path(global_step)
        model.saver.restore(sess, checkpoint_path)
        Utilities.log_checkpoint_load(checkpoint_path)

    loss = 0
    train_iterations = global_step
    while train_iterations < Config.train_max_iterations:
        # train
        l = model.train(sess)
        loss += l
        train_iterations += 1

        # show train loss
        if train_iterations % Config.display_intervals == 0:
            loss /= float(Config.display_intervals)
            Utilities.log_train_loss(train_iterations, loss)
            loss = 0

        # save checkpoint
        if train_iterations % Config.checkpoint_intervals == 0:
            checkpoint_path = model.saver.save(sess, Utilities.get_checkpoint_path(train_iterations))
            Utilities.log_checkpoint_save(checkpoint_path)


if __name__ == '__main__':
    main()
```
images, labels are queues from tf.train.shuffle_batch function. I used Graph.finalize() and I am sure no new operation added to graph in training because the size of stored models' files are equal. I guess the reason of leak is tensorflow queues.

### Environment info
Ubuntu Server 14.04.5 LTS 

Installed version of CUDA and cuDNN: 
CUDA V8.0.44, cuDNN V5.1.5.

installed tensorflow from PyPI.
1. sudo pip install --upgrade tensorflow_gpu.
2. tensorflow 0.12.1.

**Comment by yaroslavvb** [comment]
It can be useful to do a memory profile using google [performance tools](http://goog-perftools.sourceforge.net/)

**Comment by Mahdizade** [comment]
@yaroslavvb I use google-pprof and heap profile save in /tmp folder. it takes too long time to train so the profile only contain near 150 itrations. the files uploaded [here](https://www.dropbox.com/s/2r2pqs4l0bjar9h/tmp.zip?dl=1)

**Comment by yaroslavvb** [comment]
if you combine it with pprof, you can get a graph and track down which function is responsible for allocating extra memory

**Comment by Mahdizade** [comment]
It doesn't show any function, this is the output of top30 command in pprof:
```
(pprof) top30
Total: 658.2 MB
   194.3  29.5%  29.5%    194.3  29.5% 00007f8a8afae90e
   172.2  26.2%  55.7%    172.2  26.2% 00007f8a8aedbae3
    84.1  12.8%  68.5%     84.1  12.8% Eigen::internal::TensorExecutor::run::{lambda#1}@266d900
    50.2   7.6%  76.1%     50.2   7.6% 00007f8a8aec2ffe
    25.9   3.9%  80.0%     25.9   3.9% 00007f8a8af5377d
    23.2   3.5%  83.5%     23.4   3.6% PyDict_Clear
    14.2   2.2%  85.7%     14.2   2.2% 00007f8a8aec2778
    10.5   1.6%  87.3%     10.5   1.6% 00007f8a8adef4c8
     8.1   1.2%  88.5%      8.1   1.2% 00007f8a8af826d2
     6.8   1.0%  89.6%      6.8   1.0% 00007f8a8af542a3
     6.8   1.0%  90.6%      6.8   1.0% 00007f8a8af52fed
     5.7   0.9%  91.4%    596.5  90.6% extra_lbits
     4.2   0.6%  92.1%      4.2   0.6% 00007f8a8afad614
     3.7   0.6%  92.7%      3.7   0.6% 00007f8ace0b7249
     3.1   0.5%  93.1%      6.1   0.9% PyMem_Malloc
     3.1   0.5%  93.6%      4.1   0.6% ecp_nistz256_precomputed
     2.8   0.4%  94.0%      9.1   1.4% PyEval_EvalFrameEx
     2.8   0.4%  94.4%      8.6   1.3% std::_Maybe_get_result_type@26ca280
     2.4   0.4%  94.8%      2.5   0.4% PyBuffer_FillInfo
     2.1   0.3%  95.1%      2.1   0.3% __GI___strdup
     2.1   0.3%  95.4%      2.1   0.3% 00007f8a8adee7a7
     2.0   0.3%  95.7%      2.3   0.4% __FRAME_END__
     1.6   0.2%  96.0%      1.6   0.2% 00007f8a8af53762
     1.3   0.2%  96.2%      1.3   0.2% std::_Bind@26bb1c0
     1.2   0.2%  96.4%      1.2   0.2% 00007f8a8af433a1
     1.2   0.2%  96.6%      1.3   0.2% std::_Bind@26c08c0
     1.1   0.2%  96.7%      1.1   0.2% PyObject_RichCompareBool
     1.0   0.2%  96.9%      1.8   0.3% std::_Weak_result_type@26a88a0
     1.0   0.2%  97.0%      1.0   0.2% 00007f8a8aff9f73
     1.0   0.2%  97.2%      1.0   0.2% 00007f8a8b01a2ea
```
graph file uploaded [here](https://www.dropbox.com/s/mjo4kc17micbfw6/pprof3105.0.ps?dl=1), unfortunately I cant find a function is responsible for allocating extra memory

**Comment by yaroslavvb** [comment]
That usage doesn't seem very high, why do you think there's a memory leak in queue? Memory usage can increase between versions without having a leak

**Comment by Mahdizade** [comment]
because I see a 32 GB memory usage after 700000 iterations. I think the cause is data, I use tfrecord and tensorflow queue for data handling. I am sure that I hadn't leak in version 0.9.

**Comment by yaroslavvb** [comment]
@zffchen78 do you have any idea why google perf tools shows "29.5% 00007f8a8afae90e" instead of symbol name in the memory dump? Does it mean this memory was allocated outside of tensorflow?

**Comment by drpngx** [comment]
@yaroslavvb the hex address shows up when the symbol lookup failed. This is always very brittle. You can try to fiddle with build options to see if it helps. You need the frame pointers for instance, IIRC. If you gdb attach to it, then info symbol, then you might be able to reverse map it.

**Comment by carltonwang** [comment]
I got exactly the same problem as you, I also use TFRecord and Queue to process data, and my main memory keeps growing until the system goes down...
By the way, I use Tensorflow 0.12.1, CUDA 8.0.44, CUDNN 5.1, CentOS 7

**Comment by drpngx** [comment]
Main memory as in CPU? Could you build with tcmalloc, then get a heap profile?

**Comment by yaroslavvb** [comment]
To enable tcmalloc

```
sudo apt-get install google-perftools
export LD_PRELOAD="/usr/lib/libtcmalloc.so.4" 

```

**Comment by carltonwang** [comment]
I was training a seq-to-seq network like "Hierarchical Recurrent Encoder-Decoder": https://arxiv.org/abs/1507.02221
My code runs very well on CPU, the performance is also good, except that it is slower than GPU...
But when I run my code on GPU (GTX 1070, CUDA 8.0.44, CUDNN 5.1), the main memory will keep growing. 
I am using CentOS 7, can I solve this problem by gperftools? If so, how can I install and config it on CentOS? 
Thanks...

**Comment by carltonwang** [comment]
Can this problem be solved in the next release?

**Comment by yaroslavvb** [comment]
@carltonwang Could you give a reproducible example and profile as a picture/graph?
cc @ebrevdo in case he has ideas

**Comment by ebrevdo** [comment]
Are you using dynamic or static rnn?

**Comment by Mahdizade** [comment]
I use static rnn.

**Comment by carltonwang** [comment]
I use dynamic_rnn, my model is very complicated, this is the code I use to update parameters:
`VARIABLES = tf.trainable_variables()
 GRADIENTS = tf.gradients(ys=MEAN_LOSS, xs=VARIABLES)
 CLIPPED_GRADIENTS, _ = tf.clip_by_global_norm(t_list=GRADIENTS, clip_norm=gradient_clip_norm)
 OPTIMIZER = tf.train.AdamOptimizer(LEARNING_RATE)
 GRADIENT_VARIABLE_PAIRS = zip(CLIPPED_GRADIENTS, VARIABLES)
 UPDATE = OPTIMIZER.apply_gradients(grads_and_vars=GRADIENT_VARIABLE_PAIRS, global_step=GLOBAL_STEP)`
There is a very complex process before getting the MEAN_LOSS, and the model includes a hierarchical structure of RNNs in three levels... I am considering how to give you a simplified version of my code...

**Comment by carltonwang** [comment]
This is the simplified version of my code and data (in TFRecord), please run it on GPU...
My system is: CentOS 7, GTX 1070, CUDA 8.0.44, CUDNN 5.1, NVIDIA Driver 375.26

[test.zip](https://github.com/tensorflow/tensorflow/files/752806/test.zip)

**Comment by carltonwang** [comment]
Any idea about this problem?

**Comment by yaroslavvb** [comment]
@carltonwang can you isolate your problem a bit more? IE, does it happen with latest version? Does it happen with tcmalloc? Can you run it with memory profiler and figure out who allocates all the extra memory?

**Comment by carltonwang** [comment]
It happens on the latest version, with or without tcmalloc...

**Comment by carltonwang** [comment]
I implemented an encoding algorithm in my code. Today I just encapsulated this encoding method as an RNN cell, and then the memory leak disappeared!!!

**Comment by girving** [comment]
Closing since the problem is gone.  Happy to reopen if there is still an underlying bug to be fixed in TensorFlow, but it sounds like everything so far is inconclusive.

**Comment by benwu232** [comment]
@Mahdizade I met similar problem. 
My program is similar to yours and after several epoch, 32G memory exhausted. I debugged the program and found each time the saver.save() function is run, the python process got a little bigger. I recommend you try this and see if the same thing happens.

**Comment by erogol** [comment]
I have the same issue here. It goes overflowing after couple of iterations on version 1.0.1

**Comment by Mahdizade** [comment]
@benwu232 Thanks for your comment. My issue is already exists. I will test your solution but I see the leak on my validation code. till now I cant find the problem. in my validation code leak happened only in the first load of graph parameters, after validate one checkpoint and load another checkpoint, I don't see any leak in htop. I will investigate more on this problem. I don't thing the saver is the cause, I see the steady leak on training, I save the checkpoint every 5000 iteration but leak exists in this intervals.

**Comment by carltonwang** [comment]
Look at this code:
`result = tf.contrib.layers.bias_add(inputs=tf.matmul(a=left, b=right), activation_fn=tf.nn.softmax)`
If I use "softmax" as the activation function, there will be a memory leak, but if I change the activation function to "relu" or "softplus" or "sigmoid", the memory leak will disappear.
By the way, this "softmax" is not the one used at the last layer for classification.

Platform: CentOS 7, x86_64
TensorFlow version: 1.0_gpu

**Comment by aselle** [comment]
Could you provide a complete test case that exhibits this behavior so we can reproduce this?

**Comment by carltonwang** [comment]
After I update from TF 1.0.0 to TF 1.0.1, this problem disappear...

**Comment by Mahdizade** [comment]
@aselle I use tensorflow unit test for ctc_loss in [ctc_loss_op_test](https://github.com/tensorflow/tensorflow/blob/099ef355adf86296c4a9a67ed553c6d2036f99d5/tensorflow/python/kernel_tests/ctc_loss_op_test.py). I apply some changes and reorganized code for run a continuous session. The final code is here
[ctc_loss.txt](https://github.com/tensorflow/tensorflow/files/879019/ctc_loss.txt)

simply rename the file to ctc_loss.py and run it with python ctc_loss.py. I see the leak in the htop.

**Comment by artgrigoryan** [comment]
Seems memory leak here. `sess.close()`, `tf.reset_default_graph()`, `gc.collect()`  don't free memory.

    def test():
        
        x = tf.placeholder(tf.float32, [None, 28, 28])
        y_ = tf.placeholder(tf.int32, [None])
        training = tf.placeholder(tf.bool)
        
        input_layer = tf.reshape(x, [-1, 28, 28, 1])
        
        conv1 = tf.layers.conv2d(inputs=input_layer, filters=20, kernel_size=[5,5], activation=tf.nn.relu)
        pool1 = tf.layers.max_pooling2d(inputs=conv1, pool_size=[2, 2], strides = [2, 2])
        conv2 = tf.layers.conv2d(inputs=pool1, filters=40, kernel_size=[5,5], activation=tf.nn.relu)
        pool2 = tf.layers.max_pooling2d(inputs=conv2, pool_size=[2,2], strides=[2,2])
        pool2_flat = tf.reshape(pool2, [-1, 4*4*40])
        
        dense1 = tf.layers.dense(pool2_flat, units=1000, activation=tf.nn.relu)
        dropout1 = tf.layers.dropout(inputs=dense1, rate=0.5, training=training)
        dense2 = tf.layers.dense(dropout1, units=1000, activation=tf.nn.relu)
        dropout2 = tf.layers.dropout(inputs=dense2, rate=0.5, training=training)
        y = tf.layers.dense(dropout2, 10)

        softmax_layer = tf.nn.softmax(y)
        
        labels = tf.stack([tf.range(0, tf.shape(y_)[0]), y_], axis=1)
        log_likelihood = tf.reduce_mean(-tf.log(tf.gather_nd(softmax_layer, labels)))
        train_step = tf.train.GradientDescentOptimizer(0.05).minimize(log_likelihood)
        
        predict = tf.to_int32(tf.argmax(softmax_layer, 1))
        correct_prediction = tf.equal(predict, y_)
        accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        
        init = tf.global_variables_initializer()
        
        test_variable = np.random.rand(50000, 28, 28)
        validation_images, validation_labels = np.random.rand(500, 28, 28), np.random.randint(0, 10, 500)
        
        del test_variable
        
        sess = tf.Session()
        sess.run(init)
        sess.graph.finalize()
        cur_accuracy = sess.run(accuracy, feed_dict={x: validation_images, y_: validation_labels, training: False})
        
        sess.close()
        tf.reset_default_graph()
        gc.collect()

memory_profiler output:

    Line #    Mem usage    Increment   Line Contents
    ================================================
         8    115.6 MiB      0.0 MiB   def test():
         9                                 
        10    119.9 MiB      4.2 MiB       x = tf.placeholder(tf.float32, [None, 28, 28])
        11    119.9 MiB      0.0 MiB       y_ = tf.placeholder(tf.int32, [None])
        12    119.9 MiB      0.0 MiB       training = tf.placeholder(tf.bool)
        13                                 
        14    120.0 MiB      0.1 MiB       input_layer = tf.reshape(x, [-1, 28, 28, 1])
        15                                 
        16    120.3 MiB      0.3 MiB       conv1 = tf.layers.conv2d(inputs=input_layer, filters=20, kernel_size=[5,5], activation=tf.nn.relu)
        17    120.3 MiB      0.0 MiB       pool1 = tf.layers.max_pooling2d(inputs=conv1, pool_size=[2, 2], strides = [2, 2])
        18    120.9 MiB      0.6 MiB       conv2 = tf.layers.conv2d(inputs=pool1, filters=40, kernel_size=[5,5], activation=tf.nn.relu)
        19    120.9 MiB      0.0 MiB       pool2 = tf.layers.max_pooling2d(inputs=conv2, pool_size=[2,2], strides=[2,2])
        20    120.9 MiB      0.0 MiB       pool2_flat = tf.reshape(pool2, [-1, 4*4*40])
        21                                 
        22    120.9 MiB      0.0 MiB       dense1 = tf.layers.dense(pool2_flat, units=1000, activation=tf.nn.relu)
        23    120.9 MiB      0.0 MiB       dropout1 = tf.layers.dropout(inputs=dense1, rate=0.5, training=training)
        24    121.2 MiB      0.3 MiB       dense2 = tf.layers.dense(dropout1, units=1000, activation=tf.nn.relu)
        25    121.3 MiB      0.1 MiB       dropout2 = tf.layers.dropout(inputs=dense2, rate=0.5, training=training)
        26    121.3 MiB      0.0 MiB       y = tf.layers.dense(dropout2, 10)
        27                             
        28    121.3 MiB      0.0 MiB       softmax_layer = tf.nn.softmax(y)
        29                                 
        30    121.3 MiB      0.0 MiB       labels = tf.stack([tf.range(0, tf.shape(y_)[0]), y_], axis=1)
        31    121.3 MiB      0.0 MiB       log_likelihood = tf.reduce_mean(-tf.log(tf.gather_nd(softmax_layer, labels)))
        32    123.1 MiB      1.8 MiB       train_step = tf.train.GradientDescentOptimizer(0.05).minimize(log_likelihood)
        33                                 
        34    123.1 MiB      0.0 MiB       predict = tf.to_int32(tf.argmax(softmax_layer, 1))
        35    123.1 MiB      0.0 MiB       correct_prediction = tf.equal(predict, y_)
        36    123.1 MiB      0.0 MiB       accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
        37                                 
        38    123.1 MiB      0.0 MiB       init = tf.global_variables_initializer()
        39                                 
        40    422.3 MiB    299.2 MiB       test_variable = np.random.rand(50000, 28, 28)
        41    425.4 MiB      3.1 MiB       validation_images, validation_labels = np.random.rand(500, 28, 28), np.random.randint(0, 10, 500)
        42                                 
        43    126.4 MiB   -299.0 MiB       del test_variable
        44                                 
        45    248.3 MiB    121.9 MiB       sess = tf.Session()
        46    423.1 MiB    174.8 MiB       sess.run(init)
        47    423.1 MiB      0.0 MiB       sess.graph.finalize()
        48    837.2 MiB    414.1 MiB       cur_accuracy = sess.run(accuracy, feed_dict={x: validation_images, y_: validation_labels, training: False})
        49                                 
        50    837.2 MiB      0.0 MiB       sess.close()
        51    837.2 MiB      0.0 MiB       tf.reset_default_graph()
        52    837.2 MiB      0.0 MiB       gc.collect()

    
Environment: Ubuntu 16.04.2 LTS, tensorflow_gpu 1.0.1, cuda 8.0.44, cudnn 5.1.

**Comment by gunan** [comment]
looks like this was resolved with 1.0 release. Closing the issue.
